contents/linux/shell.md

```diff
@@ -0,0 +1,180 @@
+
+# check if file exists.
+test -z $filename && echo "You MUST input a filename." && exit 0
+test ! -e $filename && echo "The filename '$filename' DO NOT EXIST" && exit 0
+
+# check file type.
+test -f $filename && filetype="regular file"
+test -d $filename && filetype="directory"
+
+# check file permission.
+test -r $filename && perm="readable"
+test -w $filename && perm="$perm writable"
+test -x $filename && perm="$perm executable"
+
+# output result.
+echo "The filename: $filename is a $filetype"
+echo "And permissions are : $perm"
+```
+
+The above bash script is still just a collection of bash commands. Although it saves us a lot of works if it is a daily routine task, it still can be done by typing those commands line by line. What makes bash script really exciting is the semantics of `conditionals`, `loops` and `functions`. Which makes bash script [**turning complete**](https://en.wikipedia.org/wiki/Turing_completeness).
+
+The following is an example of bash script that make use of conditionals and loops. Suppose you want to let user input a directory name, check if it exists, and then output permissions for all files inside that directory.
+
+```bash
+#!bin/bash
+
+# read user input and check if directory exists.
+read -p "Please enter a directory name: " dir
+if [ "$dir" == "" -o ! -d "$dir" ]; then
+	echo "The $dir is NOT exist in your system"
+	exit 1
+fi
+
+# output permissions for each of the file under the directory.
+filelist=$(ls $dir)
+for filename in $filelist
+do 
+	perm=""
+	test -r "$dir/$filename" && perm="readable"
+	test -w "$dir/$filename" && perm="$perm writable"
+	test -x "$dir/$filename" && perm="$perm executable"
+	echo "The file $dir/$filename 's permission is $perm" 
+done
+```
+**Functions** in bash script just like functions in normal programming languages, more semantics of bash script can be found [here](https://www.gnu.org/software/bash/manual/html_node/). 
+
+bash script allows us to automate the frequently performed operations, it is easy to use and can be executed on any Unix-like operating systems without any modifications.
+
+### Benefit: Data Stream Redirection
+
+In linux, everything is abstracted as file, including **input**, **output**, and **error** streams. **Redirection** is a feature in Linux such that when executing a command/user program, you can change the standard input/output devices by changing its [file descriptor](https://en.wikipedia.org/wiki/File_descriptor). Each command and user program in linux is an application with input, output and errors. We can redirect the output of one application to the input of another, as such, we combined the two applications together as if we build a pipe between them.
+
+![file-descriptor](images/file-descriptor-illustration.jpg)
+
+For example, suppose you just finished your coding assignment, the resulting executable `calculate` takes a user input, do the calculation, and gives the output. To run it, you simply types:
```

Comment:
Interesting, here you use "for example" but above it was "The following is an example of..." 
:P 
Would replace do with "performs"